pkg/ccl/changefeedccl/changefeed_stmt.go

```diff
@@ -318,21 +319,78 @@ func changefeedPlanHook(
 			}
 
 			recordPTSMetricsTime := sliMetrics.Timers.PTSCreate.Start()
-
+			// ptr is the feed-level protected timestamp record which exists when per-table protected
+			// timestamps are disabled.
 			var ptr *ptpb.Record
+			// perTablePTSRecords is the per-table protected timestamp records which exist when
+			// per-table protected timestamps are enabled.
+			var perTablePTSRecords []*ptpb.Record
+			// ptsRecords is the protected timestamp records object containing all per-table protected
+			// timestamp records. Its format matches what will be persisted to the job info table.
+			var ptsRecords *cdcprogresspb.ProtectedTimestampRecords
 			codec := p.ExecCfg().Codec
-			ptr = createProtectedTimestampRecord(
-				ctx,
-				codec,
-				jobID,
-				targets,
-				details.StatementTime,
-			)
-			progress.GetChangefeed().ProtectedTimestampRecord = ptr.ID.GetUUID()
 
+			// We do not yet have the progress config here, so we need to check the settings directly.
+			perTableTrackingEnabled := changefeedbase.TrackPerTableProgress.Get(&p.ExecCfg().Settings.SV)
+			perTableProtectedTimestampsEnabled := changefeedbase.PerTableProtectedTimestamps.Get(&p.ExecCfg().Settings.SV)
+			usingPerTablePTS := perTableTrackingEnabled && perTableProtectedTimestampsEnabled
+			if usingPerTablePTS {
+				protectedTimestampRecords := make(map[descpb.ID]uuid.UUID)
+				if err := targets.EachTarget(func(target changefeedbase.Target) error {
+					ptsTargets := changefeedbase.Targets{}
+					ptsTargets.Add(target)
+					ptsRecord := createProtectedTimestampRecord(
+						ctx,
+						codec,
+						jobID,
+						ptsTargets,
+						details.StatementTime,
+					)
+					perTablePTSRecords = append(perTablePTSRecords, ptsRecord)
+					uuid := ptsRecord.ID.GetUUID()
+					protectedTimestampRecords[target.DescID] = uuid
+					return nil
+				}); err != nil {
+					return err
+				}
+				ptsRecords = &cdcprogresspb.ProtectedTimestampRecords{
+					ProtectedTimestampRecords: protectedTimestampRecords,
+				}
+			} else {
+				ptr = createProtectedTimestampRecord(
+					ctx,
+					codec,
+					jobID,
+					targets,
+					details.StatementTime,
+				)
+				progress.GetChangefeed().ProtectedTimestampRecord = ptr.ID.GetUUID()
+			}
 			jr.Progress = *progress.GetChangefeed()
 
 			if changefeedStmt.CreatedByInfo != nil {
+				// We protect the PTS records that we created earlier. There should either be a
+				// feed-level PTS record or per-table PTS record, but not both.
+				if ptr != nil {
+					pts := p.ExecCfg().ProtectedTimestampProvider.WithTxn(p.InternalSQLTxn())
+					if err := pts.Protect(ctx, ptr); err != nil {
+						return err
+					}
+				}
+				if len(perTablePTSRecords) > 0 {
```

Comment:
nit: you don't need to do a len check here, iterating over an empty slice will be a no-op

Reply:
Hmm ok I see, yeah just keep it the way it currently is then. Thanks for trying anyway!